presentation/src/main/java/com/anytypeio/anytype/presentation/sets/ViewerLayoutWidgetUi.kt

```diff
@@ -59,54 +64,123 @@ data class ViewerLayoutWidgetUi(
-    val cover = when {
-        coverRelationKey.isNullOrBlank() -> ViewerLayoutWidgetUi.State.ImagePreview.None
-        coverRelationKey == Relations.PAGE_COVER -> ViewerLayoutWidgetUi.State.ImagePreview.Cover
-        else -> {
-            val preview = storeOfRelations.getByKey(coverRelationKey)
-            if (preview != null) {
-                ViewerLayoutWidgetUi.State.ImagePreview.Custom(preview.name.orEmpty())
-            } else {
-                ViewerLayoutWidgetUi.State.ImagePreview.None
-            }
-        }
-    }
     return this.copy(
         viewer = viewer.id,
         layoutType = viewer.type,
         withIcon = ViewerLayoutWidgetUi.State.Toggle.WithIcon(!viewer.hideIcon),
         fitImage = ViewerLayoutWidgetUi.State.Toggle.FitImage(viewer.coverFit),
         cardSize = cardSize,
-        cover = cover,
-        showCardSize = showCardSize
+        showCardSize = showCardSize,
+        imagePreviewItems = viewer.getImagePreviewItems(
+            storeOfRelations = storeOfRelations,
+            relationLinks = relationLinks
+        ),
     )
 }
+
+private suspend fun DVViewer.getImagePreviewItems(
+    storeOfRelations: StoreOfRelations,
+    relationLinks: List<RelationLink>
+): List<ViewerLayoutWidgetUi.State.ImagePreview> {
+    val selectedCoverRelationKey = coverRelationKey
+
+    val validFileRelations = getValidFileRelations(storeOfRelations, relationLinks)
+
+    return buildList {
+        addNoneImagePreview(selectedCoverRelationKey)
+        addPageCoverImagePreview(selectedCoverRelationKey)
+        addCustomImagePreviews(validFileRelations, selectedCoverRelationKey)
+    }
+}
+
+private suspend fun getValidFileRelations(
+    storeOfRelations: StoreOfRelations,
+    relationLinks: List<RelationLink>
+): List<ObjectWrapper.Relation> {
+    return relationLinks
+        .filter { it.format == RelationFormat.FILE }
+        .mapNotNull { storeOfRelations.getByKey(it.key) }
+        .filter { relation ->
+            relation.isValid && relation.isHidden != true && relation.isArchived != true &&
+                    !relation.isReadonlyValue && !relation.key.isSystemKey()
+        }
+}
+
+private fun MutableList<ViewerLayoutWidgetUi.State.ImagePreview>.addNoneImagePreview(
+    selectedCoverRelationKey: String?
+) {
+    add(
+        ViewerLayoutWidgetUi.State.ImagePreview.None(
+            isChecked = selectedCoverRelationKey.isNullOrBlank()
+        )
+    )
+}
+
+private fun MutableList<ViewerLayoutWidgetUi.State.ImagePreview>.addPageCoverImagePreview(
+    selectedCoverRelationKey: String?
+) {
+    add(
+        ViewerLayoutWidgetUi.State.ImagePreview.PageCover(
+            isChecked = selectedCoverRelationKey == Relations.PAGE_COVER
+        )
+    )
+}
+
+private fun MutableList<ViewerLayoutWidgetUi.State.ImagePreview>.addCustomImagePreviews(
+    validFileRelations: List<ObjectWrapper.Relation>,
+    selectedCoverRelationKey: String?
+) {
+    validFileRelations.forEach { fileRelation ->
```

Comment:
this function doesn't look clean because of `forEach` block. why not `map` + `addAll` for instance?

Reply:
fixed